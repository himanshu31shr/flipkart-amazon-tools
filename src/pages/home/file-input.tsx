import React, { ChangeEvent, useId } from "react";
import { CheckCircle, CloudUpload } from "@mui/icons-material";
import { Button } from "@mui/material";

type FileInputProp = { 
  name: string; 
  selected: boolean;
  fileCount?: number;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  accepts: string;
  multiple?: boolean;
};

export const FileInput = (props: FileInputProp) => {
  const inputId = useId();
  
  const getButtonLabel = () => {
    if (!props.selected) {
      return `Select ${props.name}`;
    }
    
    if (props.fileCount && props.fileCount > 1) {
      return `${props.fileCount} Files Selected`;
    }
    
    return 'File Selected';
  };

  return (
    <>
      <input
        id={inputId}
        accept={props.accepts}
        style={{ display: 'none' }}
        type="file"
        onChange={props.onChange}
        multiple={props.multiple}
      />
      <label htmlFor={inputId}>
        <Button
          component="span"
          color={props.selected ? 'success' : 'primary'}
          variant={props.selected ? "contained" : "outlined"}
          tabIndex={-1}
          startIcon={props.selected ? <CheckCircle /> : <CloudUpload />}
          sx={{
            transition: 'all 0.2s ease-in-out',
            minWidth: 180,
            '&:hover': {
              transform: 'translateY(-2px)',
            }
          }}
        >
          {getButtonLabel()}
        </Button>
      </label>
    </>
  );
};
