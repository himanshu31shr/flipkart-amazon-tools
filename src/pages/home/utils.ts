import html2pdf from 'html2pdf.js';
import { MouseEvent } from "react";

export const readFileFromInput = async (file?: File): Promise<Uint8Array | null> => {
  if (!file) return null;

  const fileReader = new FileReader();

  // Wrap the file reading process in a promise to handle asynchronous operation
  return await new Promise<Uint8Array>((resolve, reject) => {
    fileReader.onload = () => {
      resolve(new Uint8Array(fileReader.result as ArrayBuffer));
    };
    fileReader.onerror = reject;
    fileReader.readAsArrayBuffer(file);
  });
};

export function downloadFile(e: MouseEvent<HTMLAnchorElement>, finalPdf: string) {
  const a = document.createElement("a");
  document.body.appendChild(a);
  a.style.display = 'none';
  a.href = finalPdf;
  a.target = "_blank";
  a.download = "pdf-merged-" + Date.now();
  a.click();
  document.body.removeChild(a);
  e.preventDefault();
  return;
}

export const exportTableToPDF = (elementId: string) => {
  const element = document.getElementById(elementId);
  if (!element) return;

  const tableStyles = {
    backgroundColor: '#ffffff',
    padding: '8px 16px',
    color: '#000000',
  };

  // Apply temporary styles
  element.style.backgroundColor = tableStyles.backgroundColor;
  const cells = [...element.getElementsByTagName('td'), ...element.getElementsByTagName('th')];
  cells.forEach(cell => {
    cell.style.padding = tableStyles.padding;
    cell.style.backgroundColor = tableStyles.backgroundColor;
    cell.style.color = tableStyles.color;
  });

  const worker = html2pdf();
  
  worker
    .set({
      margin: 0,
      filename: "summary-table.pdf",
      image: { type: "png", quality: 0.98 },
      html2canvas: { scale: 1 },
      jsPDF: { unit: "in", format: "a4", orientation: "portrait" },
    })
    .from(element)
    .save();

  // Reset styles after a delay
  setTimeout(() => {
    element.style.backgroundColor = '';
    cells.forEach(cell => {
      cell.style.padding = '';
      cell.style.backgroundColor = '';
      cell.style.color = '';
    });
  }, 1000);
};
