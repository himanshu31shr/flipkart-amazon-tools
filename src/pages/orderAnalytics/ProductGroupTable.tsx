import React, { useState } from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
  IconButton,
  TableSortLabel,
  TableFooter,
  Typography,
} from "@mui/material";
import { KeyboardArrowDown, KeyboardArrowUp } from "@mui/icons-material";
import { SkuTable } from "./SkuTable";
import { Group } from "./types/group.type";

type ProductGroupTableProps = {
  groupedData: Group[];
  openRow: string | null;
  handleRowClick: (group: string) => void;
};

type Order = "asc" | "desc";

export const ProductGroupTable: React.FC<ProductGroupTableProps> = ({
  groupedData,
  openRow,
  handleRowClick,
}) => {
  const [order, setOrder] = useState<Order>("asc");
  const [orderBy, setOrderBy] = useState<keyof Group>("group");

  const handleRequestSort = (property: keyof Group) => {
    const isAsc = orderBy === property && order === "asc";
    setOrder(isAsc ? "desc" : "asc");
    setOrderBy(property);
  };

  const sortedData = [...groupedData].sort((a, b) => {
    if (orderBy === "group") {
      return order === "asc"
        ? a.group.localeCompare(b.group)
        : b.group.localeCompare(a.group);
    } else {
      return order === "asc"
        ? Number(a[orderBy]) - Number(b[orderBy])
        : Number(b[orderBy]) - Number(a[orderBy]);
    }
  });

  return (
    <TableContainer component={Paper}>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell />
            <TableCell>
              <TableSortLabel
                active={orderBy === "group"}
                direction={orderBy === "group" ? order : "asc"}
                onClick={() => handleRequestSort("group")}
              >
                Group
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "orders"}
                direction={orderBy === "orders" ? order : "asc"}
                onClick={() => handleRequestSort("orders")}
              >
                Orders
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {sortedData.map((group) => (
            <React.Fragment key={group.group}>
              <TableRow onClick={() => handleRowClick(group.group)}>
                <TableCell>
                  <IconButton>
                    {openRow === group.group ? (
                      <KeyboardArrowUp />
                    ) : (
                      <KeyboardArrowDown />
                    )}
                  </IconButton>
                </TableCell>
                <TableCell>
                  <strong>{group.group.replaceAll("_", " ")}</strong>
                </TableCell>
                <TableCell>{group.orders}</TableCell>
                <TableCell>{group.returns}</TableCell>
                <TableCell>{group.returns}</TableCell>
                <TableCell>{group.returns}</TableCell>
                <TableCell>{group.returns}</TableCell>
              </TableRow>
              <SkuTable skus={group.skus} open={openRow === group.group} />
            </React.Fragment>
          ))}
        </TableBody>
        <TableFooter>
          <TableRow>
            <TableCell colSpan={2}>
              <Typography variant="h6" gutterBottom component="div">
                Totals
              </Typography>
            </TableCell>
            <TableCell>
              <Typography variant="h6" gutterBottom component="div">
                {sortedData.reduce((acc, curr) => acc + curr.orders, 0)}
              </Typography>
            </TableCell>
            <TableCell>
              <Typography variant="h6" gutterBottom component="div">
                {sortedData.reduce((acc, curr) => acc + curr.returns, 0)}
              </Typography>
            </TableCell>
          </TableRow>
        </TableFooter>
      </Table>
    </TableContainer>
  );
};
