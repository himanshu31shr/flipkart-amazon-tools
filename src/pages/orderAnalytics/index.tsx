import React, { useEffect, useState } from "react";
import {
  Box,
  Container,
  Grid,
  Paper,
  Typography,
  CircularProgress,
  IconButton,
  Badge,
  Button,
} from "@mui/material";
import FilterListIcon from "@mui/icons-material/FilterList";
import CalendarTodayIcon from "@mui/icons-material/CalendarToday";
import { format } from "date-fns";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../../store";
import { fetchAllOrdersForAnalytics } from "../../store/slices/allOrdersForAnalyticsSlice";
import { fetchProducts } from "../../store/slices/productsSlice";
import { fetchCategories } from "../../store/slices/categoriesSlice";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFns";
import CategoryDistributionTable from "./components/CategoryDistributionTable";
import CategoryOrdersChart from "./components/CategoryOrdersChart";
import OrderMetrics from "./components/OrderMetrics";
import SkuOrdersChart from "./components/SkuOrdersChart";
import TopProductsTable from "./components/TopProductsTable";
import DateRangeFilter from "./components/DateRangeFilter";
import FilterPopover from "./components/FilterPopover";
import { useOrderFilters } from "./hooks/useOrderFilters";

const OrderAnalytics: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const {
    items: allOrders,
    loading: allOrdersLoading,
    error: allOrdersError,
  } = useSelector((state: RootState) => state.allOrdersForAnalytics);
  const {
    items: products,
    loading: productsLoading,
    error: productsError,
  } = useSelector((state: RootState) => state.products);
  const {
    items: categories,
    loading: categoriesLoading,
    error: categoriesError,
  } = useSelector((state: RootState) => state.categories);

  // State for popovers
  const [filterAnchorEl, setFilterAnchorEl] = useState<null | HTMLElement>(
    null
  );
  const [dateAnchorEl, setDateAnchorEl] = useState<null | HTMLElement>(null);

  // Use custom hook for filters
  const {
    filterState,
    skuOptions,
    platformOptions,
    productOptions,
    filteredOrders,
    updateFilter,
    clearFilters,
  } = useOrderFilters({ allOrders, categories, products });

  useEffect(() => {
    if (allOrders.length === 0 && !allOrdersLoading && !allOrdersError) {
      dispatch(fetchAllOrdersForAnalytics());
    }
    if (products.length === 0 && !productsLoading && !productsError) {
      dispatch(fetchProducts({}));
    }
    if (categories.length === 0 && !categoriesLoading && !categoriesError) {
      dispatch(fetchCategories());
    }
  }, []);

  const isLoading = allOrdersLoading || productsLoading || categoriesLoading;
  const error = allOrdersError || productsError || categoriesError;

  // Popover handlers
  const handleFilterClick = (event: React.MouseEvent<HTMLElement>) => {
    setFilterAnchorEl(event.currentTarget);
  };
  const handleFilterClose = () => {
    setFilterAnchorEl(null);
  };
  const handleDateClick = (event: React.MouseEvent<HTMLElement>) => {
    setDateAnchorEl(event.currentTarget);
  };
  const handleDateClose = () => {
    setDateAnchorEl(null);
  };

  // Active filter count
  const activeFilterCount = [
    filterState.selectedCategory,
    filterState.selectedSku,
    filterState.selectedPlatform,
    filterState.selectedProduct,
  ].filter(Boolean).length;

  // Date range summary
  const dateSummary = `${format(
    filterState.dateRange.startDate,
    "dd MMM yyyy"
  )} - ${format(filterState.dateRange.endDate, "dd MMM yyyy")}`;

  if (error) {
    return (
      <Container maxWidth="xl" sx={{ mt: 4, mb: 4 }}>
        <Typography color="error" variant="h6">
          Error: {error}
        </Typography>
      </Container>
    );
  }

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <Container maxWidth="xl" sx={{ mt: 4, mb: 4 }}>
        <Box
          display="flex"
          justifyContent="space-between"
          alignItems="center"
          mb={3}
        >
          <Typography variant="h4">Order Analytics</Typography>
          <Box display="flex" gap={2}>
            <Button
              variant="outlined"
              startIcon={<CalendarTodayIcon />}
              onClick={handleDateClick}
              sx={{ textTransform: "none" }}
            >
              {dateSummary}
            </Button>
            <IconButton
              onClick={handleFilterClick}
              color={activeFilterCount > 0 ? "primary" : "default"}
              sx={{ ml: 1 }}
            >
              <Badge badgeContent={activeFilterCount} color="primary">
                <FilterListIcon />
              </Badge>
            </IconButton>
          </Box>
        </Box>

        <DateRangeFilter
          dateRange={filterState.dateRange}
          onDateRangeChange={(startDate, endDate) =>
            updateFilter("dateRange", { startDate, endDate })
          }
          anchorEl={dateAnchorEl}
          onClose={handleDateClose}
        />

        <FilterPopover
          anchorEl={filterAnchorEl}
          onClose={handleFilterClose}
          categories={categories}
          selectedCategory={filterState.selectedCategory}
          selectedSku={filterState.selectedSku}
          selectedPlatform={filterState.selectedPlatform}
          selectedProduct={filterState.selectedProduct}
          skuOptions={skuOptions}
          platformOptions={platformOptions}
          productOptions={productOptions}
          onCategoryChange={(category) =>
            updateFilter("selectedCategory", category)
          }
          onSkuChange={(sku) => updateFilter("selectedSku", sku)}
          onPlatformChange={(platform) =>
            updateFilter("selectedPlatform", platform)
          }
          onProductChange={(product) =>
            updateFilter("selectedProduct", product)
          }
          onClearFilters={clearFilters}
        />

        {isLoading ? (
          <Box
            display="flex"
            justifyContent="center"
            alignItems="center"
            minHeight="400px"
          >
            <CircularProgress />
          </Box>
        ) : (
          <>
            <OrderMetrics
              orders={filteredOrders}
              products={products}
              categories={categories}
            />
            <Paper sx={{ p: 2 }}>
              <Grid container spacing={3}>
                <Grid item xs={12} md={7}>
                  <Typography variant="h6" gutterBottom>
                    Category Distribution Details
                  </Typography>
                  <CategoryDistributionTable
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
                <Grid item xs={12} md={5}>
                  <Typography variant="h6" gutterBottom>
                    Orders by Category
                  </Typography>
                  <CategoryOrdersChart
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
              </Grid>
            </Paper>
            <Paper sx={{ p: 2, mt:4, height: "100%" }}>
              <Grid container spacing={3}>
                <Grid item xs={12} md={7}>
                  <Typography variant="h6" gutterBottom>
                    Top Products
                  </Typography>
                  <TopProductsTable
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
                <Grid item xs={12} md={5}>
                  <Typography variant="h6" gutterBottom>
                    Orders by SKU
                  </Typography>
                  <SkuOrdersChart
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
              </Grid>
            </Paper>
          </>
        )}
      </Container>
    </LocalizationProvider>
  );
};

export default OrderAnalytics;

/* TODO: Create the CategoryDistributionChart component */
