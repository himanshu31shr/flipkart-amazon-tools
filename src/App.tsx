import { Download, MergeOutlined } from "@mui/icons-material";
import { Alert, Button } from "@mui/material";
import Box from "@mui/material/Box";
import Container from "@mui/material/Container";
import Grid from "@mui/material/Grid2";
import { ChangeEvent, useState } from "react";
import { FileInput } from "./components/file-input";
import { mergePdfs } from "./services/merge.service";
import { downloadFile, readFileFromInput } from "./utils";
import { AppBar } from "./components/appbar";

export default function App() {
  const [amazon, setAmazon] = useState<File>();
  const [flipkart, setFlipkart] = useState<File>();
  const [finalPdf, setFinal] = useState<string>();

  const handleSubmit = async () => {
    if (!amazon && !flipkart) return;

    const pdf = await mergePdfs({
      amzon: await readFileFromInput(amazon),
      flp: await readFileFromInput(flipkart),
    });

    const outputPdfBytes = await pdf.save();
    const blob = new Blob([outputPdfBytes], { type: "application/pdf" });

    // Create a Blob URL for the PDF
    const pdfUrl = URL.createObjectURL(blob);
    setFinal(pdfUrl);
    setAmazon(undefined);
    setFlipkart(undefined);
  };

  return (
    <>
      <AppBar />
      <Container maxWidth="sm">
        <Alert sx={{ my: 4 }} severity="info">
          Please select AMAZON and FLIPKART labels correctly to process and
          merge the PDFs
        </Alert>

        <Grid container spacing={2} sx={{ my: 4 }}>
          <Grid>
            <Box>
              <FileInput
                name="amazon"
                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                  if (e.target.files) {
                    setAmazon(e.target.files[0]);
                  }
                }}
                selected={!!amazon}
              />
            </Box>
          </Grid>
          <Grid>
            <Box>
              <FileInput
                name="flipkart"
                selected={!!flipkart}
                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                  if (e.target.files) {
                    setFlipkart(e.target?.files[0]);
                  }
                }}
              />
            </Box>
          </Grid>
        </Grid>
        <Box sx={{ my: 4 }}>
          <Button
            component="label"
            variant="contained"
            tabIndex={-1}
            onClick={() => {
              handleSubmit();
            }}
            startIcon={<MergeOutlined />}
            disabled={!amazon && !flipkart}
          >
            Merge & Generate Pdf
          </Button>
        </Box>
        {finalPdf && (
          <>
            <Box sx={{ my: 4 }}>
              <Button
                component="label"
                variant="contained"
                tabIndex={-1}
                href={finalPdf}
                startIcon={<Download />}
                onClick={() => {
                  downloadFile(finalPdf);
                }}
              >
                Download Pdf
              </Button>
            </Box>
            <iframe src={finalPdf} width={"100%"} height={"800"}></iframe>
          </>
        )}
      </Container>
    </>
  );
}
