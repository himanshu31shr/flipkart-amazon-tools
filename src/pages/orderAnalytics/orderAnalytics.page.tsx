import CloseIcon from "@mui/icons-material/Delete";
import { Alert, Box, CircularProgress, IconButton } from "@mui/material";
import Grid from "@mui/material/Grid2";
import Papa from "papaparse";
import React, { useEffect, useState } from "react";
import { deleteDatabase, getOrders } from "../../storage/db";
import { FileInput } from "../home/file-input";
import { ProductGroupTable } from "./ProductGroupTable";
import { OrderAggregationService } from "./services/orderAggregation.service";
import { Group } from "./types/group.type";
import { OrderItem } from "./types/order.type";

export const OrderAnalyticsPage = () => {
  const [orderItems, setOrderItems] = useState<OrderItem[]>([]);
  const [groupedData, setGroupedData] = useState<Group[]>([]);
  const [timeline, setTimeLine] = useState<
    Partial<{ start: string; end: string }>
  >({});

  const [fetchingDb, setFetchingDb] = useState(false);

  const fetchOrders = async () => {
    setFetchingDb(true);
    const savedOrders = await getOrders();
    if (savedOrders.length > 0) {
      setOrderItems(savedOrders);
      const service = new OrderAggregationService(savedOrders as OrderItem[]);
      setGroupedData(await service.process());
      setTimeLine(service.getDates());
    }
    setFetchingDb(false);
  };

  const deleteDb = () => {
    setFetchingDb(true);
    deleteDatabase();
    setOrderItems([]);
    setGroupedData([]);
    setTimeLine({});
    setFetchingDb(false);
  };

  useEffect(() => {
    fetchOrders();
  }, []);

  const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFetchingDb(true);
    const file = event.target.files?.[0];
    if (file) {
      Papa.parse(file, {
        complete: async (results: Papa.ParseResult<OrderItem>) => {
          const data = results.data as OrderItem[];
          setOrderItems(data);
          const service = new OrderAggregationService(data);
          setGroupedData(await service.process());
          setTimeLine(service.getDates());
          setFetchingDb(false);
        },
        header: true,
        transform: (value: string, field: string) => {
          if (field === "sku") {
            value = value.replaceAll(/["]|sku:/gi, "");
          }

          return value;
        },
        skipEmptyLines: true,
      });
    }
  };

  return (
    <Loader showLoader={fetchingDb}>
      <Grid
        container
        spacing={2}
        justifyContent={"space-between"}
        marginTop={2}
        marginBottom={2}
      >
        <Box>
          <FileInput
            onChange={handleFileUpload}
            name="Flipkart Report"
            accepts=".csv"
            selected={orderItems == null}
          />
        </Box>
        <Box>
          <IconButton onClick={() => deleteDb()} data-testid="delete-data">
            <CloseIcon />{" "}
          </IconButton>
        </Box>
        <Grid size={12}>
          {timeline.start && timeline.end && (
            <Alert color="info">
              Showing data between {new Date(timeline.start).toDateString()} &{" "}
              {new Date(timeline.end).toDateString()}
            </Alert>
          )}
        </Grid>
      </Grid>
      {groupedData.length > 0 && (
        <ProductGroupTable
          groupedData={groupedData}
        />
      )}
    </Loader>
  );
};

const Loader = ({
  showLoader = false,
  children,
}: {
  showLoader: boolean;
  children: React.ReactNode;
}) => {
  return (
    <>
      {showLoader && (
        <Grid spacing={10} marginTop={2} alignItems={"center"}>
          <CircularProgress size={16} />{" "}Building data...
        </Grid>
      )}
      {children}
    </>
  );
};
