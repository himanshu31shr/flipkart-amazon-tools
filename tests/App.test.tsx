import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import "@testing-library/jest-dom";
import App from "../src/App";
import { HomePage } from "../src/pages/home/home.page";
import { OrderAnalyticsPage } from "../src/pages/orderAnalytics/orderAnalytics.page";

test("renders label merger link", () => {
  const { container } = render(<App />);
  const linkElement = screen.getByText(/label merger/i);
  expect(linkElement).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("renders navigation links", () => {
  render(<App />);

  const menu = screen.getByTestId("menu-button");
  expect(menu).toBeInTheDocument();
  fireEvent.click(menu);
  const mergeLabelsLink = screen.getByText(/merge labels/i);
  const analyticsLink = screen.getByText(/analytics/i);
  expect(mergeLabelsLink).toBeInTheDocument();
  expect(analyticsLink).toBeInTheDocument();
});

test("routing works correctly", async () => {
  render(<App />);

  const menu = screen.getByTestId("menu-button");
  expect(menu).toBeInTheDocument();
  fireEvent.click(menu);
  const mergeLabelsLink = screen.getByText(/merge labels/i);
  const analyticsLink = screen.getByText(/analytics/i);
  expect(mergeLabelsLink).toBeInTheDocument();
  expect(analyticsLink).toBeInTheDocument();

  fireEvent.click(mergeLabelsLink);
  const homePageElement = await screen.findByText(
    /Please select AMAZON and FLIPKART labels correctly to process and merge the PDFs/gi
  ); // Adjust this based on actual content
  expect(homePageElement).toBeInTheDocument();

  fireEvent.click(menu);
  fireEvent.click(analyticsLink);
  const analyticsPageElement = await screen.findByText(/Flipkart/gi); // Adjust this based on actual content
  expect(analyticsPageElement).toBeInTheDocument();
});

test("renders HomePage component when navigating to /labelMerger/", async () => {
  render(
    <MemoryRouter initialEntries={["/labelMerger/"]}>
      <Routes>
        <Route path="/labelMerger/" element={<HomePage />} />
      </Routes>
    </MemoryRouter>
  );
  const homePageElement = await screen.findByText(
    /Please select AMAZON and FLIPKART labels correctly to process and merge the PDFs/gi
  ); // Adjust this based on actual content
  expect(homePageElement).toBeInTheDocument();
});

test("renders OrderAnalyticsPage component when navigating to /labelMerger/analytics/", async () => {
  render(
    <MemoryRouter initialEntries={["/labelMerger/analytics/"]}>
      <Routes>
        <Route
          path="/labelMerger/analytics/"
          element={<OrderAnalyticsPage />}
        />
      </Routes>
    </MemoryRouter>
  );
  const analyticsPageElement = await screen.findByText(/Flipkart/gi); // Adjust this based on actual content
  expect(analyticsPageElement).toBeInTheDocument();
});
