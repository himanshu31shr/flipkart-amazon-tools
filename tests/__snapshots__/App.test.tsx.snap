// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders label merger link 1`] = `
<div>
  <div
    class="MuiBox-root css-k008qs"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1343boy-MuiPaper-root-MuiAppBar-root"
      style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1ygil4i-MuiToolbar-root"
      >
        <button
          aria-label="menu"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-xxdts0-MuiButtonBase-root-MuiIconButton-root"
          data-testid="menu-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
            />
          </svg>
        </button>
        <button
          aria-label="close drawer"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-ly16f5-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="ChevronLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </button>
        <div
          class="MuiTypography-root MuiTypography-h6 css-1bv5yfe-MuiTypography-root"
        >
          Label Merger
        </div>
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-rehya9-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-161ms7l-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              aria-label="dark mode toggle"
              class="PrivateSwitchBase-input MuiSwitch-input css-j8yymo"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-17jyosd-MuiSwitch-thumb"
            />
          </span>
          <span
            class="MuiSwitch-track css-tj8d99-MuiSwitch-track"
          />
        </span>
      </div>
    </header>
    <div
      class="MuiDrawer-root MuiDrawer-docked css-1ufa4ms-MuiDrawer-docked"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-1lwhjos-MuiPaper-root-MuiDrawer-paper"
        style="--Paper-shadow: none;"
      >
        <div
          class="MuiBox-root css-1hskriy"
          role="presentation"
        >
          <div
            class="css-1kh1mwg"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-5lnhk4-MuiTypography-root"
            >
              Sacred Sutra
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="ChevronLeftIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
          />
          <ul
            class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
              data-discover="true"
              data-testid="merge-labels"
              href="/labelMerger/"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              >
                <div
                  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                      data-testid="MoveToInboxIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 3H4.99c-1.11 0-1.98.9-1.98 2L3 19c0 1.1.88 2 1.99 2H19c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 12h-4c0 1.66-1.35 3-3 3s-3-1.34-3-3H4.99V5H19zm-3-5h-2V7h-4v3H8l4 4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                    >
                      Merge Labels
                    </span>
                  </div>
                </div>
              </li>
            </a>
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
              data-discover="true"
              data-testid="transactions"
              href="/labelMerger/transactions/"
            >
              <li
                class="MuiListItem-root MuiListItem-gutters css-19f6boz-MuiListItem-root"
              >
                <div
                  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cokf1l-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                      data-testid="ReceiptIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 17H6v-2h12zm0-4H6v-2h12zm0-4H6V7h12zM3 22l1.5-1.5L6 22l1.5-1.5L9 22l1.5-1.5L12 22l1.5-1.5L15 22l1.5-1.5L18 22l1.5-1.5L21 22V2l-1.5 1.5L18 2l-1.5 1.5L15 2l-1.5 1.5L12 2l-1.5 1.5L9 2 7.5 3.5 6 2 4.5 3.5 3 2z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
                    >
                      Transaction Analytics
                    </span>
                  </div>
                </div>
              </li>
            </a>
          </ul>
        </div>
      </div>
    </div>
    <main
      class="css-8vfclp"
    >
      <div
        class="MuiContainer-root css-mmdczc-MuiContainer-root"
      />
    </main>
  </div>
</div>
`;
